### DISCUSSION TITLE
Use of I2C IMU Sensor

### DISCUSSION THREAD

**Discussion by DrCoconut1245** [mikaelnousiainen/RS41ng#51] [Ideas]
Hi, I have been looking into using the RS41 as a low cost telemetry unit for model rockets. I have already tested it once on a flight where it preformed well and we where able to get live GPS data for the majority of the flight, but I have been wondering about the possibility of connecting a IMU over I2C and having that data transmitted back as well.  What would the steps be to configure a new sensor to work on the sonde? Is this a approachable task for a beginner programmer? I have no idea if I will or even can do this, but I want to get an understanding of the amount of work this would take. Thank you for any advice!
[Sample IMU](https://www.adafruit.com/product/4464)

**Comment by mikaelnousiainen** [comment]
@DrCoconut1245 Apologies for a late reply.

The IMU you've linked works on 3.3V voltage, so it should work with RS41 hardware.

There are the following high-level steps to getting your sensor working, but I'm afraid it is not a task for a beginner:

1. Add driver code to read the sensor readings from the chip using I²C commands. There is an example of how to structure your code in the BMP280 sensor code: https://github.com/mikaelnousiainen/RS41ng/blob/main/src/drivers/bmp280/bmp280.c
    -  Add a new directory under `src/drivers` and the driver code there.
    - Usually there is some C code available for any sensor (search for Arduino drivers for the IMU, for example) that can be used as a basis and ported to the STM32 hardware APIs.

2. Add calls to `telemetry.c` to read the sensor: https://github.com/mikaelnousiainen/RS41ng/blob/main/src/telemetry.c#L24 and some new fields to the `telemetry_data` struct to store the readings.

3. Add new template variables in `template.c` (https://github.com/mikaelnousiainen/RS41ng/blob/main/src/template.c) so that the new sensor readings can be referenced in APRS/CW and other messages (in `config.c`)

4. If you wish to include the readings in Horus 4FSK transmission custom fields, customize Horus 4FSK v2 packets here: https://github.com/mikaelnousiainen/RS41ng/blob/main/src/codecs/horus/horus_packet_v2.c#L55

**Reply by DrCoconut1245** [reply]
Awesome, thank you so much! This does seem a bit complicated and probably out of my abilities, but if I do ever approach this (or more accurately ask someone else if they can help me do this) I will use this as a starting point. Thank you very much!

**Reply by mikaelnousiainen** [reply]
Ok!

Is the sample IMU link you gave the exact model you'd be using? I could take a look at this at some point (no promises though), but I'd need to buy this sensor too.

Some comments:

- The GPS mode may need to be adjusted if the acceleration is high (over 1G), as by default the firmware configures the GPS in airborne mode for less than 1G acceleration. You can configure GPS for up to 4G acceleration by changing the dynModel (dynamic model platform) value, see: https://github.com/mikaelnousiainen/RS41ng/blob/main/src/drivers/ubxg6010/ubxg6010.c#L457

- Is the data rate (how often the sonde sends APRS or Horus) enough for your use case? You cannot get access to second- or millisecond level variations in sensor readings at all. There's also not much memory available for buffering readings for delayed transmission. All I'm saying is that while an RS41 probably works as a general-purpose tracker, you might need something more capable for more accurate tracking of rocket flights.

**Reply by DrCoconut1245** [reply]
The sensor I chose was just an example, I didnt do much research beyond finding a chip that looked like it would work. I could find a actual chip if you do want to program it. I didn't know that there was a setting that stopped GPS transmissions at high g's, and that would explain why we only got GPS data near the apogee of the flight and when it landed. I thought it would be limited by the GPS chip itself, which was willing to operate far past what the rocket would have put it through. I will have to change that for the next flight. I am not overly concerned about the data rate, as adding this chip would just make profiling the flight easier. I wanted to try using the sonde because most other methods of GPS tracking model rockets costs around $50-$100, where as if I lose a sonde I can just find another one. Plus, its way cooler than just buying something off the shelf.
